panc/src/main/java/org/quattor/pan/dml/functions/Join.java

```diff
@@ -0,0 +1,127 @@
+        if (operations[0] instanceof Null || operations[1] instanceof Null) {
+            throw SyntaxException.create(sourceRange, MSG_VALUE_CANNOT_BE_NULL, "join");
+        }
+
+        // If there are only two arguments, the second argument needs to be a list.
+        if (operations.length == 2) {
+            if (operations[1] instanceof Element) {
+                if (!(operations[1] instanceof ListResource)) {
+                    throw SyntaxException.create(sourceRange, MSG_SECOND_ARG_LIST_OR_VARIABLE_REF, "join");
+                }
+            }
+        }
+
+        return new Join(sourceRange, operations);
+    }
+
+    @Override
+    public Element execute(Context context) throws EvaluationException {
+
+        assert (ops.length != 1);
+
+        // Extract the first argument. This must be a string value.
+        String delimeter = null;
+        try {
+            delimeter = ((StringProperty) ops[0].execute(context)).getValue();
+        } catch (ClassCastException cce) {
+            throw new EvaluationException("first argument in join() must be a string", getSourceRange(), context);
+        }
+
+        // Create a list containing all elements we need to join.
+        ArrayList<String> result = new ArrayList<String>();
+
+        // If the second argument is a ListResource, add the strings to the result-list.
+        if (ops.length == 2) {
+
+            // Double check whether the second argument is actually a list.
+            Element list = ops[1].execute(context);
+
+            if (list instanceof ListResource) {
+
+                // Loop over the list items and convert them to strings.
+                Resource.Iterator it = ((ListResource) list).iterator();
+                while (it.hasNext()) {
+                    Element e = it.next().getValue();
+
+                    if (e instanceof Resource) {
+                        throw new EvaluationException(
+                                "join() doesn't accept nested elements in the passed list",
+                                getSourceRange(), context);
+                    }
+
+                    if (!(e instanceof StringProperty)) {
+                        throw new EvaluationException(
+                                "join() only accepts strings as elements in the list",
+                                getSourceRange(), context);
+                    }
+
+                    result.add(((StringProperty) e).getValue());
+                }
+
+            } else {
+                throw new EvaluationException(
+                        "join() only accepts strings or a list of strings as input arguments",
+                        getSourceRange(), context);
+            }
+
+        // If all arguments are passed individually, add them to a list.
+        } else {
+            int length = ops.length;
+
+            for (int i = 1; i < length; i++) {
+                Element e = ops[i].execute(context);
```

Comment:
It's not recursive, it just loops over the arguments and executes them. The execution is needed to get the actual value when for example passing variables to the function.